fem/transfer.hpp

```diff
@@ -244,6 +244,103 @@ class L2ProjectionGridTransfer : public GridTransfer
    virtual const Operator &BackwardOperator();
 };
 
+
+/** @brief Transfer data in H1 finite element spaces between a coarse mesh and an
+    embedded refined mesh using L2 projection. */
+/** The forward, coarse-to-fine, transfer uses L2 projection. The backward,
+    fine-to-coarse, transfer is defined as B = (F^t M_f F)^{-1} F^t M_f, where
+    F is the forward transfer matrix, and M_f is the mass matrix lumped through
+    row-summation. Note that the backward transfer operator, B, is a left
+    inverse of the forward transfer operator, F, i.e. B F = I. Both F and B are
+    defined in physical space.
+
+    This class supports H1 finite element spaces and fine meshes that are a
+    uniform refinement of the coarse mesh.  Generally, the coarse and fine FE
+    spaces can have different orders, however, in order for the backward
+    operator to be well-defined, the number of fine dofs (in a coarse element)
+    should not be smaller than the number of coarse dofs. */
+class L2ProjectionH1GridTransfer : public GridTransfer
+{
+protected:
+   /** Class representing projection operator between a high-order H1 finite
+       element space on a coarse mesh, and a low-order H1 finite element space
+       on a refined mesh (LOR). We assume that the low-order space, fes_lor,
+       lives on a mesh obtained by refining the mesh of the high-order space,
+       fes_ho. */
+   class L2ProjectionH1 : public Operator
+   {
+      const FiniteElementSpace& fes_ho;
+      const FiniteElementSpace& fes_lor;
+
+      SparseMatrix* R;
+      SparseMatrix M_LH;
+      mutable DenseMatrix* P;
+      const double p_rtol;
+      const double p_atol;
+
+      Table ho2lor;
+   public:
+      L2ProjectionH1(const FiniteElementSpace& fes_ho_,
+                     const FiniteElementSpace& fes_lor_);
+      virtual ~L2ProjectionH1();
+      /// Perform the L2 projection onto the LOR space
```

Comment:
@ebchin It might be worth clarifying the intended use/purpose of the functions below in terms of the HO->LO or LO->HO for both primal and dual fields. I recall Mult() being a forward map (HO->LO) for primal fields, but MultTranspose() the backward (LO->HO) map for dual fields. I feel like this type of thing could lead to some confusion when trying to be exercised at a higher level. The fact that there is language like Forward, Backward, Restriction, Prolongation, and the use of these flips depending on primal or dual field transfer makes me think some clarifying "use" comments might go a long way. Let me know if this comment doesn't make sense.

Reply:
I agree, more extensive comments here would be beneficial.  I added a more complete description of the functions.  Hopefully these new comments clear it up?

Reply:
👍